cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -1,195 +1,213 @@
- * LDS response names all RDS routes in a snapshot, to ensure that Envoy makes the request for all EDS clusters or RDS
- * routes eventually.
+ * {@code SimpleCache} provides a default implementation of {@link SnapshotCache}. It maintains a single versioned
+ * {@link Snapshot} per node group. For the protocol to work correctly in ADS mode, EDS/RDS requests are responded to
+ * only when all resources in the snapshot xDS response are named as part of the request. It is expected that the CDS
+ * response names all EDS clusters, and the LDS response names all RDS routes in a snapshot, to ensure that Envoy makes
+ * the request for all EDS clusters or RDS routes eventually.
  *
- * <p>SimpleCache can also be used as a config cache for distinct xDS requests. The snapshots are required to contain
- * only the responses for the particular type of the xDS service that the cache serves. Synchronization of multiple
- * caches for different response types is left to the configuration producer.
+ * <p>The snapshot can be partial, e.g. only include RDS or EDS resources.
  */
-public class SimpleCache<T> implements Cache<T> {
+public class SimpleCache<T> implements SnapshotCache<T> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(SimpleCache.class);
 
-  private final Consumer<T> callback;
+  private final boolean ads;
   private final NodeGroup<T> groups;
-  private final ExecutorService executorService;
 
-  private final Object lock = new Object();
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private final Lock readLock = lock.readLock();
+  private final Lock writeLock = lock.writeLock();
 
   @GuardedBy("lock")
   private final Map<T, Snapshot> snapshots = new HashMap<>();
   @GuardedBy("lock")
-  private final Table<T, Long, Watch> watches = HashBasedTable.create();
+  private final Map<T, CacheStatusInfo> statuses = new HashMap<>();
 
   @GuardedBy("lock")
   private long watchCount;
 
-  /**
-   * Constructs a simple cache that uses {@link ForkJoinPool#commonPool()} to execute async callbacks.
-   *
-   * @param callback callback invoked when a group is seen for the first time
-   * @param groups maps an envoy host to a node group
-   */
-  public SimpleCache(Consumer<T> callback, NodeGroup<T> groups) {
-    this(callback, groups, ForkJoinPool.commonPool());
-  }
-
   /**
    * Constructs a simple cache.
    *
-   * @param callback callback invoked when a group is seen for the first time
+   * @param ads if running in ADS-mode, responses are delayed until all resources are explicitly named
    * @param groups maps an envoy host to a node group
-   * @param executorService executor service used to execute async callbacks
    */
-  public SimpleCache(Consumer<T> callback, NodeGroup<T> groups, ExecutorService executorService) {
-    this.callback = callback;
+  public SimpleCache(boolean ads, NodeGroup<T> groups) {
```

Comment:
it'd be a little bit surprising as a consumer of this api that I need to tell the cache that it's going to be used for ADS... i understand the why because I've read the code, but i'd be moderately annoyed if i found myself having to maintain multiple identical caches if i want to serve both ADS and non-ADS. 

doing anything about that is obviously out of the scope of this change, just voicing my opinion

Reply:
Totally agree. We should create a follow-up issue to do some refactoring so that the `DiscoveryServer` is able to communicate which mode it should be operating in. Off the top of my head, I suspect we could do that by making the `ads` flag part of the `Watch` itself, so that when the server creates the watch it can toggle the mode automatically.